File: solutions/318/main.cpp
/* ************************************************************************
> File Name:     main.cpp
> Author:        Sean
> Project:    LeetCode
> Created Time:  Tue Nov 16 15:21:58 2021
> Description:   
 ************************************************************************/
#include <vector> 
#include <string> 
#include <gtest/gtest.h> 

using namespace std; 

class Solution {
public:
    int maxProduct(vector<string>& words) {
        int result = 0;
        vector<int> wordsInt(words.size(), 0);
        for (int i = 0; i < words.size(); ++i) {
            string w = words[i];
            for (const auto& c : w) {
                wordsInt[i] |= 1 << (c - 'a');
            }
        }
        for (int i = 0; i < words.size(); ++i) {
            for (int j = i+1; j < words.size(); ++j) {
                if (wordsInt[i] & wordsInt[j]) continue;
                else {
                    if (words[i].size() * words[j].size() > result)
                        result = words[i].size() * words[j].size();
                }
            }
        }
        return result;
    }
};

TEST(Test318, SimpleTest) {
    vector<string> words{
        "abcw","baz","foo","bar","xtfn","abcdef"
    };
    Solution s; 
    EXPECT_EQ(s.maxProduct(words), 16);
}

int main() {
    testing::InitGoogleTest();
    return RUN_ALL_TESTS();
}

File: solutions/318/main.py
from typing import List 

class Solution:
    def maxProduct(self, words: List[str]) -> int:
        maxProduct = 0
        wordsInt = []
        for w in words:
            result = 0
            for c in w:
                result |= 1 << ord(c) - ord('a')
            wordsInt.append(result)
        for i in range(len(words)):
            for j in range(i+1, len(words)):
                if wordsInt[i] & wordsInt[j]:
                    continue
                else:
                    product = len(words[i]) * len(words[j])
                    if product > maxProduct:
                        maxProduct = product
        return maxProduct

if __name__ == "__main__": 
    words = ["abcw","baz","foo","bar","xtfn","abcdef"]
    s = Solution()
    assert s.maxProduct(words) == 16
